share/functions/__fish_complete_ant_targets.fish

```diff
@@ -1,16 +1,54 @@
 function __fish_complete_ant_targets -d "Print list of targets from build.xml and imported files"
-    set -l buildfile "build.xml"
-    if test -f $buildfile
-        # show ant targets
-        __fish_filter_ant_targets $buildfile
-
-        # find files with buildfile
-        set files (sed -n "s/^.*<import[^>]* file=[\"']\([^\"']*\)[\"'].*\$/\1/p" < $buildfile)
+    function __get_buildfile -d "Get a buildfile that will be used by ant"
+        set -l tokens $argv # tokens from 'commandline -co'
+        set -l prev $tokens[1]
+        set -l buildfile "build.xml"
+        for token in $argv[2..-1]
+            switch $prev
+                case -buildfile -file -f
+                    set buildfile (eval echo $token)
+            end
+            set prev $token
+        end
+        # return last one
+        echo $buildfile
+    end
+    function __parse_ant_targets_from_projecthelp -d "Parse ant targets from projecthelp"
+        set -l buildfile $argv[1] # full path to buildfile
+        set -l targets (ant -p -debug -f $buildfile 2> /dev/null | string match -r '^\s\S+.*$' $projecthelp)
+        for target in $targets
+            set -l tokens (string match -r '^\s([[:graph:]]+)(?:\s+([[:print:]]+))?' "$target")
+            if [ (count $tokens) -ge 3 ]
+                echo $tokens[2]\t$tokens[3]
+            else if [ (count $tokens) -ge 2 ]
+                echo $tokens[2]
+            end
+        end
+    end
+    function __get_ant_targets_from_projecthelp -d "Get ant targets from projecthelp"
+        set -l buildfile $argv[1] # full path to buildfile
 
-        # iterate through files and display their targets
-        for file in $files
+        set -l cache_dir
+        if [ \( -n $__fish_user_data_dir \) -a \( -d $__fish_user_data_dir \) ]
+            set cache_dir $__fish_user_data_dir/ant_completions
+        else
+            set cache_dir "$HOME/.local/share/fish/ant_completions"
+        end
+        mkdir -p $cache_dir
 
-            __fish_filter_ant_targets $file
+        set -l cache_file $cache_dir/(string escape --style=var $buildfile)
```

Comment:
Though [gradle's completion](https://github.com/fish-shell/fish-shell/blob/master/share/completions/gradle.fish#L48) uses `md5` hash to name cache files, I adopt escaping so that we can easily guess where actual files are.

Reply:
So there are a few things to consider here. One is the ease of figuring out which file is which, as you say. The other is that many empty nested directories are expensive on some platforms, and make cache invalidation take a lot longer than a single-level hashed names implementation would.

It's your call, but I personally would use file names while developing my code and then switch it over to md5 when it's production ready.

Reply:
I will use `md5` hash. Thanks.